File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
def solution(record):
    answer = []
    res = []
    dic = {} # userId 별로 닉네임 저장할 공간
    for i in record:
        tmp = i.split() # 명령, userId, nickname 순
        if tmp[0] == "Enter":
            dic[tmp[1]] = tmp[2]
            answer.append((tmp[1],tmp[0]))
        elif tmp[0] == "Leave":
            answer.append((tmp[1],tmp[0]))
        else:
            dic[tmp[1]] = tmp[2]
        
    for ans in answer:
        if ans[1] =="Enter":
            res.append(dic[ans[0]]+"님이 들어왔습니다.")
        else:
            res.append(dic[ans[0]]+"님이 나갔습니다.")
    
    return res

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;
class Solution {
    public String[] solution(String[] record) {
        ArrayList<String> answer = new ArrayList<>();
        HashMap<String, String> rMap = new HashMap<>();
        
        for(String str : record){
            String[] part = str.split(" ");
            if(part[0].equals("Leave")) continue;
            rMap.put(part[1], part[2]);
        }
        
        for(String str : record){
            String[] part = str.split(" ");
            switch(part[0]) {
                case "Enter":
                    answer.add(rMap.get(part[1])+"님이 들어왔습니다.");
                    break;
                case "Leave":
                    answer.add(rMap.get(part[1])+"님이 나갔습니다.");
                    break;
            }
        }
        
        return answer.toArray(new String[0]);
    }
}